source/azure_iot_json_writer.c

```diff
@@ -0,0 +1,465 @@
+                                                                 bool usValue )
+{
+    AzureIoTResult_t xResult;
+    az_result xCoreResult;
+    az_span xPropertyNameSpan;
+
+    if( ( pxWriter == NULL ) || ( pucPropertyName == NULL ) || ( ulPropertyNameLength == 0 ) )
+    {
+        AZLogError( ( "AzureIoTJSONWriter_AppendPropertyWithBoolValue failed: invalid argument" ) );
+        return eAzureIoTErrorInvalidArgument;
+    }
+
+    xPropertyNameSpan = az_span_create( ( uint8_t * ) pucPropertyName, ( int32_t ) ulPropertyNameLength );
+
+    if( az_result_failed( xCoreResult = az_json_writer_append_property_name( &pxWriter->_internal.xCoreWriter, xPropertyNameSpan ) ) ||
+        az_result_failed( xCoreResult = az_json_writer_append_bool( &pxWriter->_internal.xCoreWriter, usValue ) ) )
+    {
+        AZLogError( ( "Could not append property and bool: core error=0x%08x", xCoreResult ) );
+        xResult = eAzureIoTErrorFailed;
+    }
+    else
+    {
+        xResult = eAzureIoTSuccess;
+    }
+
+    return xResult;
+}
+
+AzureIoTResult_t AzureIoTJSONWriter_AppendPropertyWithStringValue( AzureIoTJSONWriter_t * pxWriter,
+                                                                   const uint8_t * pucPropertyName,
+                                                                   uint32_t ulPropertyNameLength,
+                                                                   const uint8_t * pucValue,
+                                                                   uint32_t ulValueLen )
+{
+    AzureIoTResult_t xResult;
+    az_result xCoreResult;
+    az_span xPropertyNameSpan;
+    az_span xValueSpan;
+
+    if( ( pxWriter == NULL ) || ( pucPropertyName == NULL ) || ( ulPropertyNameLength == 0 ) ||
+        ( pucValue == NULL ) || ( ulValueLen == 0 ) )
+    {
+        AZLogError( ( "AzureIoTJSONWriter_AppendPropertyWithStringValue failed: invalid argument" ) );
+        return eAzureIoTErrorInvalidArgument;
+    }
+
+    xPropertyNameSpan = az_span_create( ( uint8_t * ) pucPropertyName, ( int32_t ) ulPropertyNameLength );
+    xValueSpan = az_span_create( ( uint8_t * ) pucValue, ( int32_t ) ulValueLen );
+
+    if( az_result_failed( xCoreResult = az_json_writer_append_property_name( &pxWriter->_internal.xCoreWriter, xPropertyNameSpan ) ) ||
+        az_result_failed( xCoreResult = az_json_writer_append_string( &pxWriter->_internal.xCoreWriter, xValueSpan ) ) )
+    {
+        AZLogError( ( "Could not append property and string: core error=0x%08x", xCoreResult ) );
+        xResult = eAzureIoTErrorFailed;
+    }
+    else
+    {
+        xResult = eAzureIoTSuccess;
+    }
+
+    return xResult;
+}
+
+int32_t AzureIoTJSONWriter_GetBytesUsed( AzureIoTJSONWriter_t * pxWriter )
+{
+    if( pxWriter == NULL )
+    {
+        AZLogError( ( "AzureIoTJSONWriter_GetBytesUsed failed: invalid argument" ) );
+        return -1;
+    }
+
+    return pxWriter->_internal.xCoreWriter._internal.total_bytes_written;
```

Comment:
Fine for now I guess, but I don't like that we have to access an internal.  Shouldn't embedded SDK expose this if it's a legit useful thing?

Reply:
Yea let me double check on this I think this is lifted straight from azureRTOS. If not I can check to see if this can be made public.

Reply:
Updated